### ISSUE TITLE
Pass var to route handler

### ISSUE THREAD

**Issue by davidebettio** [gin-gonic/gin#420]
Hi,

how can I pass a var (db connection pointer) to a route handler like this:

apiV1.GET("/users/:userid", userHandler)

so that I can use in userHandler ?

Now I use a global var but I want to move the handler functions to another package

**Comment by nazwa** [comment]
You can create a database middleware and add connection to context.

```go

// This method collects all the errors and submits them to Rollbar
func Database(connString string) gin.HandlerFunc {
	db := sqlx.MustConnect("mysql", connString)
	if err := RunMigrations(db); err != nil {
		panic(err)
	}

	return func(c *gin.Context) {
		c.Set("DB", db)
		c.Next()
	}
}



r.Use(middleware.Database(config.Config.DSN))
```

**Comment by itsbalamurali** [comment]
Doesn't open multiple connections?
Dont we need to close other connections before we open new?

**Comment by nazwa** [comment]
No, all good. Note that the outer function wraps the scope with the db variable, and only the `return func` is repeated for every http connection. 

Also golang mysql handles pooling and connection management of database connections, so all good here :+1:

**Comment by top** [comment]
@nazwa I checked your code, but still don't know how to use Database middleware :(

**Comment by nazwa** [comment]
hey @top! It's super simple, really.

The middleware has two "parts". 

Part one is what is executed once, when you initalize your middleware. That's where you set up all the global objects, connections etc. Everything thta happens one per application lifetime.

Part two is what executes on every request. For a database middleware you simply inject your "global" db object into the context. Once it's inside the context, you can retrieve it from within other middlewares and your handler function.

```go

func Database(connString string) gin.HandlerFunc {
  // <----
// This is part one
// ---->

    return func(c *gin.Context) {
   // <----
   // This is part two
  // ---->
        c.Set("DB", db)
        c.Next()
    }
}

```

And inside my handler:
```go
func (this *UploadsHandler) Upload(c *gin.Context) {
    db := c.MustGet("DB").(*bolt.DB) // MustGet returns an interface so you have to type cast it first :)
```

Let me know if it's still unclear

**Comment by rskumar** [comment]
IIRC, a new connection will be open or acquired from pool for each request. What if that handler does not use any DB. But this middleware will always add one.
In that case, a lazy func call would be good.

**Comment by crossle** [comment]
How to use context.Context store DB in gin?

**Comment by Tom29** [comment]
@nazwa what's the purpose of using `Next()`? Do i need to use it after each `Set()`? I've read the comment but don't understand

**Comment by rskumar** [comment]
You may have registered many middlewares in chain. `Next()` will keep calling next middleware and your `handler` func in last. If you do not call `Next`, your request won't be processed any further.

**Comment by delphinus** [comment]
@nazwa @rskumar You do not need `c.Next()` in this case.  Because all of middlewares & handlers in the chain are called in a row automatically without `c.Next()`.

Here's the flow to handle requests in gin.

1. `net/http` calls [`ServeHTTP()` method][servehttp] in `*gin.Engine`.
2. `handleHTTPRequest()` method will be called, and [the first `c.Next()`][fstnext] is called.
3. `c.Next()` executes [`for` loop][for] to make the chain work.  This calls handlers/middlewares sequentially, so each one does not need to call `c.Next()` in itself.

[servehttp]: https://github.com/gin-gonic/gin/blob/bf7803815b0baa22ff7a10457932882dfbf09925/gin.go#L320-L329
[fstnext]: https://github.com/gin-gonic/gin/blob/bf7803815b0baa22ff7a10457932882dfbf09925/gin.go#L359
[for]: https://github.com/gin-gonic/gin/blob/bf7803815b0baa22ff7a10457932882dfbf09925/context.go#L107-L109

----

But you need to call `c.Next()` explicitly when you do something *after* the children.  Here's the example.

```go
func someMW() gin.HandlerFunc {
  return func(c *gin.Context) {
    log.Println("before calling handler")
    c.Next()
    log.Println("after calling handler")
  }
}

func someHandler() gin.HandlerFunc {
  return func(c *gin.Context) {
    log.Println("start handler")
    doSomeFantasticThings()
    log.Println("finish handler")
  }
}

r := gin.New()
r.GET("/", someMW(), someHandler())
```

This shows log below...

```
before calling handler
start handler
finish handler
after calling handler
```

You can try to remove `c.Next()` and see the order of logs changes, and  middlewares/handlers are called sequentially without `c.Next()`.

```
before calling handler
after calling handler
start handler
finish handler
```

When you want to *cut* the chain and return on ahead, you should call `c.Abort()` explicitly.  If you replace the `c.Next()` with `c.Abort()`, you will see this.


```sh
before calling handler
# it aborts here
after calling handler
# show `after` log, but it does not continue into the handler
```

**Comment by alexmatsak** [comment]
What is the difference between passing variables through middleware and custom structure like in https://github.com/gin-gonic/gin/issues/932#issuecomment-306242400.

What is the **right way** ?

**Comment by nazwa** [comment]
@alexmatsak I personally use a combination of the two. "global" things (db, logger) etc go through context, more individual items go directly into the "module" struct object.

Context is a bit more flexible because it allows you to do self-registering modules but it's all down to what you really need.

**Comment by tarasmatsyk** [comment]
Hi @nazwa, thanks for the example, could you please elaborate a bit more and tell how to make sure that connection is closed gracefully when server is shutdown?

**Comment by colossus21** [comment]
Add a custom Header in middleware then, receive the Header content in handler. Just a theory though!
